apps/dashboard/app/(app)/authorization/permissions/components/upsert-permission/index.tsx

```diff
@@ -0,0 +1,229 @@
+import { zodResolver } from "@hookform/resolvers/zod";
+import { PenWriting3, Plus } from "@unkey/icons";
+import { Button, DialogContainer, FormInput, FormTextarea } from "@unkey/ui";
+import { useEffect, useState } from "react";
+import { FormProvider } from "react-hook-form";
+import { useUpsertPermission } from "./hooks/use-upsert-permission";
+import { type PermissionFormValues, permissionSchema } from "./upsert-permission.schema";
+
+const FORM_STORAGE_KEY = "unkey_upsert_permission_form_state";
+
+type ExistingPermission = {
+  id: string;
+  name: string;
+  slug: string;
+  description?: string;
+};
+
+const getDefaultValues = (
+  existingPermission?: ExistingPermission,
+): Partial<PermissionFormValues> => {
+  if (existingPermission) {
+    return {
+      permissionId: existingPermission.id,
+      name: existingPermission.name,
+      slug: existingPermission.slug,
+      description: existingPermission.description || "",
+    };
+  }
+
+  return {
+    name: "",
+    slug: "",
+    description: "",
+  };
+};
+
+type UpsertPermissionDialogProps = {
+  existingPermission?: ExistingPermission;
+  triggerButton?: boolean;
+  isOpen?: boolean;
+  onClose?: () => void;
+};
+
+export const UpsertPermissionDialog = ({
+  existingPermission,
+  triggerButton,
+  isOpen: externalIsOpen,
+  onClose: externalOnClose,
+}: UpsertPermissionDialogProps) => {
+  const [internalIsOpen, setInternalIsOpen] = useState(false);
+  const isEditMode = Boolean(existingPermission?.id);
+
+  // Use external state if provided, otherwise use internal state
+  const isDialogOpen = externalIsOpen !== undefined ? externalIsOpen : internalIsOpen;
+  const setIsDialogOpen =
+    externalOnClose !== undefined
+      ? (open: boolean) => !open && externalOnClose()
+      : setInternalIsOpen;
+
+  const storageKey = isEditMode
+    ? `${FORM_STORAGE_KEY}_edit_${existingPermission?.id}`
+    : FORM_STORAGE_KEY;
+
+  const formId = `upsert-permission-form-${existingPermission?.id || "new"}`;
+
+  const methods = usePersistedForm<PermissionFormValues>(
+    storageKey,
+    {
+      resolver: zodResolver(permissionSchema),
+      mode: "onChange",
+      shouldFocusError: true,
+      shouldUnregister: true,
+    },
+    "memory",
+  );
+
+  const {
+    register,
+    formState: { errors, isValid },
+    handleSubmit,
+    reset,
+    clearPersistedData,
+    saveCurrentValues,
+    loadSavedValues,
+  } = methods;
+
+  useEffect(() => {
+    if (!isDialogOpen) {
+      return;
+    }
+
+    const loadData = async () => {
+      if (existingPermission) {
+        // Edit mode
+        const hasSavedData = await loadSavedValues();
+        if (!hasSavedData) {
+          const editValues = getDefaultValues(existingPermission);
+          reset(editValues);
+        }
+      } else {
+        // Create mode
+        const hasSavedData = await loadSavedValues();
+        if (!hasSavedData) {
+          reset(getDefaultValues());
+        }
+      }
+    };
```

Comment:
can you do something like: 

```
const hasSavedData = await loadSavedValues();
if (!hasSavedData) {
    reset(getDefaultValues(existingPermission))
}
```

The getDefaultValues can handle passing in a falsy `existingPermission` -- there doesn't seem to be any difference in these if/else conditions, just whether you're passing in an existingPermission if it exists or you don't.

Reply:
I did some incremental changes and probably forgot to clean up this one